src/main/java/uk/gov/hmcts/ccd/domain/service/casedataaccesscontrol/DefaultCaseDataAccessControl.java

```diff
@@ -11,75 +10,102 @@
-                                        FakeRoleAssignmentsGenerator fakeRoleAssignmentsGenerator,
+                                        PseudoRoleAssignmentsGenerator pseudoRoleAssignmentsGenerator,
                                         ApplicationParams applicationParams,
-                                        AccessProfileService accessProfileService) {
+                                        AccessProfileService accessProfileService,
+                                        PseudoRoleToAccessProfileGenerator pseudoRoleToAccessProfileGenerator) {
         this.roleAssignmentService = roleAssignmentService;
         this.securityUtils = securityUtils;
         this.caseService = caseService;
         this.caseTypeService = caseTypeService;
         this.roleAssignmentsFilteringService = roleAssignmentsFilteringService;
-        this.fakeRoleAssignmentsGenerator = fakeRoleAssignmentsGenerator;
+        this.pseudoRoleAssignmentsGenerator = pseudoRoleAssignmentsGenerator;
         this.applicationParams = applicationParams;
         this.accessProfileService = accessProfileService;
+        this.pseudoRoleToAccessProfileGenerator = pseudoRoleToAccessProfileGenerator;
     }
 
     // Returns Optional<CaseDetails>. If this is not enough think of wrapping it in a AccessControlResponse
     // that contains the additional information about the operation result
     @Override
     public Optional<CaseDetails> applyAccessControl(CaseDetails caseDetails) {
         RoleAssignments roleAssignments = roleAssignmentService.getRoleAssignments(securityUtils.getUserId());
-        CaseDetails cloned = caseService.clone(caseDetails);
 
         RoleAssignmentFilteringResult filteringResults = roleAssignmentsFilteringService
             .filter(roleAssignments, caseDetails);
 
-        if (filteringResults.hasGrantTypeExcludedRole()) {
-            filteringResults = filteringResults.retainBasicAndStandardGrantTypeRolesOnly();
-        }
-
         if (applicationParams.getEnablePseudoRoleAssignmentsGeneration()) {
-            List<RoleAssignment> augmentedRoleAssignments = fakeRoleAssignmentsGenerator
-                .addFakeRoleAssignments(filteringResults);
+            List<RoleAssignment> pseudoRoleAssignments = pseudoRoleAssignmentsGenerator
+                .createPseudoRoleAssignments(filteringResults);
+            filteringResults = augment(filteringResults, pseudoRoleAssignments);
         }
 
         CaseTypeDefinition caseTypeDefinition = caseTypeService.getCaseType(caseDetails.getCaseTypeId());
 
+        if (applicationParams.getEnablePseudoAccessProfilesGeneration()) {
+            List<RoleToAccessProfileDefinition> generatedAccessProfiles =
+                pseudoRoleToAccessProfileGenerator.generate(caseTypeDefinition);
+
+            List<RoleToAccessProfileDefinition> roleToAccessProfiles = caseTypeDefinition.getRoleToAccessProfiles();
+            roleToAccessProfiles.addAll(generatedAccessProfiles);
```

Comment:
not the best solution this one, to change the content of the CaseTypeDefinition from its original data.
I would have rather made a copy of the definition's RoleToAccessProfiles list, augmented it, and passed that as input to the next stages rather than the whole case type.
My fear is that if in the future we need to for any reason use the original list of access profiles mappings, that's been changed and "dirtied" with pseudo data. This could cause bugs.
Maybe a minor for now, but changing class contents in place is usually source of bugs

Reply:
I could pass Map<String, RoleToAccessProfileDefinition> roleToAccessProfilesMap to the AccessProfileServiceImpl, remove the map from the CaseTypeDefinition and build it in the DefaultCaseDataAccessControl client code. This would fix the 'hidden' role of CaseTypeDefinition.setRoleToAccessProfiles (comment above). Would have to change the AccessProfileService interface as well.